### PR TITLE
chore: adaptations for nightly-2024-05-11

### PR DESCRIPTION
**Pull request by kim-em**
These are mostly adaptations for https://github.com/leanprover/lean4/pull/4061.

Some are no problem: `rfl` proofs that no longer work, and we replace with proofs by the appropriate lemmas.

But I think the cases where we use `unseal` might need further attention.

### CODE REVIEW

File: Mathlib/Data/Fin/Basic.lean [hunk 1]
```diff
@@ -779,6 +780,7 @@ theorem zero_ne_one' [NeZero n] : (0 : Fin (n + 1)) ≠ 1 := Fin.ne_of_lt one_po
 
 #align fin.succ_zero_eq_one' Fin.succ_zero_eq_one
 
+unseal Nat.modCore in
```
Comment (semorrison) [inline Mathlib/Data/Fin/Basic.lean:783]: I'm not particularly happy with this. How would I discover the need to add this unseal, after seeing the `rfl` fail on `@Eq (Fin (n✝ + 1 + 1).succ) (succ 1) 2`?

File: Mathlib/GroupTheory/CommutingProbability.lean [hunk 2]
```diff
@@ -166,9 +166,11 @@ def reciprocalFactors (n : ℕ) : List ℕ :=
   else
     n % 4 * n :: reciprocalFactors (n / 4 + 1)
 
-@[simp] lemma reciprocalFactors_zero : reciprocalFactors 0 = [0] := rfl
+@[simp] lemma reciprocalFactors_zero : reciprocalFactors 0 = [0] :=  by
+  unfold reciprocalFactors; rfl
```
Comment (semorrison) [inline Mathlib/GroupTheory/CommutingProbability.lean:170]: Should these use `unseal` so they remain eligible for `dsimp`?

File: Mathlib/Data/Fin/Basic.lean (hunk 1 above)
Comment (nomeata) [reply Mathlib/Data/Fin/Basic.lean:782]: I'm not happy either. Any ideas?

We could just make `Nat.modCore` semireducible again, of course.

Or try harder to make `Nat.mod` reduce when the first argument is any fully known value and it's reducibly smaller than the second.

File: Mathlib/GroupTheory/CommutingProbability.lean (hunk 2 above)
Comment (nomeata) [reply Mathlib/GroupTheory/CommutingProbability.lean:170]: I don't have a good sense yet where we need `dsimp` compatibility. We could err on the side of more `unseal`.

Comment (semorrison) [reply Mathlib/GroupTheory/CommutingProbability.lean:170]: I guess let's leave them. I doubt this file is used outside of Mathlib yet, so if CI is happy so am I.

File: Mathlib/Data/Fin/Basic.lean (hunk 1 above)
Comment (nomeata) [reply Mathlib/Data/Fin/Basic.lean:782]: Or use a structurally recursive definition of `Nat.mod` (with fuel)

Comment (nomeata) [reply Mathlib/Data/Fin/Basic.lean:782]: Ah, I think I found the snag why I couldn’t make that work initially. I tried

```
def mod' (n m : Nat) := match n with
  | 0 => 0
  | _ => if n < m then n else Nat.modCore n m
```
but it wouldn’t reduce where I wanted it to.

I now see why. Note that
```
example : decide (m+3 ≤ 2) = false := rfl -- does not work
example : decide (2 < m+3) = false := rfl -- works
```

So 
```
def mod' (n m : Nat) := match n with
  | 0 => 0
  | _ => if m ≤ n then Nat.modCore n m else n
```
might do what we want (namely reduce when `n` is `0`, or when `n` is known and `m` is known enough.) Will try.

Comment (nomeata) [reply Mathlib/Data/Fin/Basic.lean:784]: Thanks for nudging me to try harder, I think I found the right magic incarnation to make this work:
https://github.com/leanprover/lean4/pull/4145

Comment (semorrison) [reply Mathlib/Data/Fin/Basic.lean:784]: Okay, we will wait on this PR until nightly-2024-05-14 lands containing https://github.com/leanprover/lean4/pull/4145!

**Comment by semorrison** [comment]
I'll open a new PR now that we're at nightly-2024-05-14.

**Pull request closed without merging**